# Chapter 3: Processing Raw Text

In [115]:
import nltk, re, pprint
from nltk import word_tokenize
from urllib import request

In [116]:
url = 'http://www.gutenberg.org/files/2554/2554-0.txt'
response = request.urlopen(url) 
raw = response.read().decode('utf-8')
print(type(raw), len(raw), raw[:75])

<class 'str'> 1176812 The Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky


In [117]:
tokens = word_tokenize(raw)
print(type(tokens), len(tokens), tokens[:10])

<class 'list'> 257058 ['\ufeffThe', 'Project', 'Gutenberg', 'eBook', 'of', 'Crime', 'and', 'Punishment', ',', 'by']


In [118]:
text = nltk.Text(tokens)
print(text[1024:1062])
print(text.collocations())

['insight', 'impresses', 'us', 'as', 'wisdom', '...', 'that', 'wisdom', 'of', 'the', 'heart', 'which', 'we', 'seek', 'that', 'we', 'may', 'learn', 'from', 'it', 'how', 'to', 'live', '.', 'All', 'his', 'other', 'gifts', 'came', 'to', 'him', 'from', 'nature', ',', 'this', 'he', 'won', 'for']
Katerina Ivanovna; Pyotr Petrovitch; Pulcheria Alexandrovna; Avdotya
Romanovna; Rodion Romanovitch; Marfa Petrovna; Sofya Semyonovna; old
woman; Project Gutenberg-tm; Porfiry Petrovitch; Amalia Ivanovna;
great deal; young man; Nikodim Fomitch; Project Gutenberg; Ilya
Petrovitch; Andrey Semyonovitch; Hay Market; Dmitri Prokofitch; Good
heavens
None


In [119]:
url1 = 'http://news.bbc.co.uk/2/hi/health/2284783.stm'
html = request.urlopen(url1).read().decode('utf8')
print(type(html), len(html))
print(html[:60])
#print(html)

<class 'str'> 37306
<!doctype html public "-//W3C//DTD HTML 4.0 Transitional//EN


In [120]:
from bs4 import BeautifulSoup 
raw1 = BeautifulSoup(html, 'html.parser').get_text() 
tokens_html = word_tokenize(raw1)
text_html = nltk.Text(tokens_html)
text_html.concordance('gene')
#print(tokens_html)

print(html.find('of the'))

Displaying 7 of 7 matches:
hey say too few people now carry the gene for blondes to last beyond the next 
blonde hair is caused by a recessive gene . In order for a child to have blond
 have blonde hair , it must have the gene on both sides of the family in the g
ere is a disadvantage of having that gene or by chance . They do n't disappear
des would disappear is if having the gene was a disadvantage and I do not thin
er's Polio campaign launched in Iraq Gene defect explains high blood pressure 
er's Polio campaign launched in Iraq Gene defect explains high blood pressure 
24338


In [121]:
hp_1 = open('/Users/maria.onoeva/Desktop/new_folder/HP_1.txt')
raw_HP = hp_1.read()

In [122]:
a = [1, 2, 3, 4, 5, 6, 7, 6, 5, 4, 3, 2, 1]
b = [' ' * 2 * (7 - i) + 'very' * i for i in a]
for line in b:
    print(line)

            very
          veryvery
        veryveryvery
      veryveryveryvery
    veryveryveryveryvery
  veryveryveryveryveryvery
veryveryveryveryveryveryvery
  veryveryveryveryveryvery
    veryveryveryveryvery
      veryveryveryvery
        veryveryvery
          veryvery
            very


In [123]:
my_sentence = 'It\'s the worst case scenario lullaby'
for i in my_sentence.split(' '):
    print(i)
    
for char in my_sentence: 
    print(char, end=' ')

It's
the
worst
case
scenario
lullaby
I t ' s   t h e   w o r s t   c a s e   s c e n a r i o   l u l l a b y 

In [124]:
word_list = [w for w in nltk.corpus.words.words('en') if w.islower()]
# print([w for w in word_list if re.search('ed$', w)])
# print([w for w in word_list if re.search('^..j..t..$', w)])
# print([w for w in word_list if re.search('..j..t..', w)])
# print([w for w in word_list if re.search('^[g-o]+$', w)])
# print([w for w in word_list if re.search('^[g-o]$', w)])
print([w for w in word_list if re.search('^[a-fj]+$', w)])
print([w for w in word_list if re.search('^[a-fj-o]$', w)])
print([w for w in word_list if re.search('^[a-o]$', w)])

['a', 'aa', 'aba', 'abac', 'abaca', 'abaff', 'abb', 'abed', 'acca', 'accede', 'ace', 'ad', 'adad', 'add', 'adda', 'added', 'ade', 'adead', 'ae', 'aface', 'affa', 'ajaja', 'b', 'ba', 'baa', 'baba', 'babe', 'bac', 'bacaba', 'bacca', 'baccae', 'bad', 'bade', 'bae', 'baff', 'bajada', 'be', 'bead', 'beaded', 'bebed', 'bed', 'bedad', 'bedded', 'bedead', 'bedeaf', 'bee', 'beef', 'bejade', 'c', 'ca', 'cab', 'caba', 'cabda', 'cad', 'cade', 'caeca', 'caffa', 'ce', 'cede', 'cee', 'd', 'da', 'dab', 'dabb', 'dabba', 'dace', 'dad', 'dada', 'dade', 'dae', 'daff', 'de', 'dead', 'deaf', 'deb', 'decad', 'decade', 'dee', 'deed', 'deedeed', 'deface', 'e', 'ea', 'ebb', 'ecad', 'edea', 'efface', 'f', 'fa', 'facade', 'face', 'faced', 'fad', 'fade', 'faded', 'fae', 'faff', 'fe', 'fed', 'fee', 'feed', 'j', 'jab', 'jabbed', 'jade', 'jaded', 'jed', 'jeff']
['a', 'b', 'c', 'd', 'e', 'f', 'j', 'k', 'l', 'm', 'n', 'o']
['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o']


In [125]:
print([w for w in word_list if re.search('^..j..t..$', w)])
print([w for w in word_list if re.search(r'^..j..t..$', w)])

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector', 'unjilted', 'unjolted', 'unjustly']
['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector', 'unjilted', 'unjolted', 'unjustly']


In [126]:
# https://stackoverflow.com/questions/26318287/what-does-r-mean-before-a-regex-pattern
print('\n') # Prints a newline character
print(r'\n') # Escape sequence is not processed
print('\b') # Prints a backspace character
print(r'\b') # Escape sequence is not processed



\n

\b


I guess I got it. So by prefixing with 'r' we kinda say to Python not to read into that but simply to pass a combination to re library, ok!

In [127]:
print([int(n) for n in re.findall(r'[0-9]{2,}', '2009-12-31')])

[2009, 12, 31]


In [128]:
paranoid = """Please could you stop the noise?
I'm tryna get some rest
From all the unborn chicken
Voices in my head"""
pattern = r'^[A-Z]'
pattern1 = r'\?$'
pattern2 = r'es$' # finds nothing because it thinks of the line as the whole string and no line ends with 'es', so in order it to work I have to tokenize it first 

nltk.re_show(pattern, paranoid)

{P}lease could you stop the noise?
{I}'m tryna get some rest
{F}rom all the unborn chicken
{V}oices in my head


In [129]:
from nltk.corpus import brown 
hobbies_learned = nltk.Text(brown.words(categories = ['hobbies', 'learned']))
hobbies_learned.findall(r'<\w*> <and> <other> <\w*s>')
hobbies_learned.findall(r'<as> <\w*> <as> <\w*>')

speed and other activities; water and other liquids; tomb and other
landmarks; Statues and other monuments; pearls and other jewels;
charts and other items; roads and other features; figures and other
objects; military and other areas; demands and other factors;
abstracts and other compilations; iron and other metals
as accurately as possible; as well as the; as faithfully as possible;
as much as what; as neat as a; as simple as you; as well as other; as
well as other; as involved as determining; as well as other; as
important as another; as accurately as possible; as accurate as any;
as much as any; as different as a; as Orphic as that; as coppery as
Delawares; as good as another; as large as small; as well as ease; as
well as their; as well as possible; as straight as possible; as well
as nailed; as smoothly as the; as soon as a; as well as injuries; as
well as many; as well as reason; as well as in; as well as of; as well
as a; as well as summer; as well as providing; as important a

## Exercises 
#### 1. Define a string s = 'colorless'. Write a Python statement that changes this to "colourless" using only the slice and concatenation operations.

In [130]:
s = 'colorless' 
print(s[:4] + 'u' + s[4:])

colourless


#### 2. We can use the slice notation to remove morphological endings on words. For example, 'dogs'[:-1] removes the last character of dogs, leaving dog. Use slice notation to remove the affixes from these words (we've inserted a hyphen to indicate the affix boundary, but omit this from your strings): dish-es, run-ning, nation-ality, un-do, pre-heat.

In [131]:
ex2_words = [w for w in 'dish-es, run-ning, nation-ality, un-do, pre-heat'.split(', ')]
ex2_noaff = [ex2_words[0][:-3], ex2_words[1][:-5], ex2_words[2][:-6], ex2_words[3][-2:], ex2_words[4][-4:]]
print(ex2_noaff)

['dish', 'run', 'nation', 'do', 'heat']


#### 3. We saw how we can generate an IndexError by indexing beyond the end of a string. Is it possible to construct an index that goes too far to the left, before the start of the string? 
I don't think so. I cannot use negative numbers as they remove chars and the lowest possible is then 0 which is the first char.
 
Okay, I see it now. I thought about slicing but not about indexing. So it's possible to raise the error then. 

In [132]:
s = "hello"
print(s[-6:])  # This will print 'hello'
print(s[-100:])  # Still prints 'hello', as slicing is more forgiving
print(s[-1])  # This will print 'o'
print(s[-5])  # This will print 'h'
# Accessing beyond the start (too far to the left)
#print(s[-6])  # This will raise an IndexError

hello
hello
o
h


#### 4. We can specify a "step" size for the slice. The following returns every second character within the slice: monty[6:11:2]. It also works in the reverse direction: monty[10:5:-2] Try these for yourself, then experiment with different step values.

In [133]:
print(paranoid[::2])
print(paranoid[10:5:-2])

Pes ol o tptenie
' rn e oers
rmalteubr hce
ocsi yha
lo 


#### 5. What happens if you ask the interpreter to evaluate monty[::-1]? Explain why this is a reasonable result.
It reverses a string, but why can it be reasonable? Perhaps if some text is reversed, it can reverse it back. Or to print some text so it's readable in mirrors. 

In [134]:
print(paranoid[::-1])
print(paranoid)

daeh ym ni secioV
nekcihc nrobnu eht lla morF
tser emos teg anyrt m'I
?esion eht pots uoy dluoc esaelP
Please could you stop the noise?
I'm tryna get some rest
From all the unborn chicken
Voices in my head


#### 6. Describe the class of strings matched by the following regular expressions. 
1. ```[a-zA-Z]+```: '+' means one or more of previous item, in this case one or more of any letter, case doesn't matter ('cat', 'so') > any word 
2. ```[A-Z][a-z]*```: here case matters, so first it should be an upper one then zero or more of any lower case ('D', 'Dog') > any upper case word 
3. ```p[aeiou]{,2}t```: a sequence should start with 'p' and end with 't', then between there should be at most 2 vowels > too much output, replacing it with ```^p[aeiou]{,2}t$``` ('put', 'poet') 
4. ```\d+(\.\d+)?```: one or more digits and zero or one ('?') of digits after the decimal point
5. ```([^aeiou][aeiou][^aeiou])*```: zero or more instances of a chunk that doesn't start with a vowel then has one vowel and doen't end with the vowel, so its CVC* combinations (??)  -- huh? 
6. ```\w+|[^\w\s]+```: any alphanumeric char OR something that doesn't start with any alphanumeric char and a white space -- anything that is not a space?

In [135]:
nltk.re_show(r'[a-zA-Z]+', paranoid)
nltk.re_show(r'[A-Z][a-z]*', paranoid)
print([w for w in word_list if re.search(r'^p[aeiou]{,2}t$', w)])
print([w for w in word_list if re.search(r'\d+(\.\d+)?', w)])
nltk.re_show(r'([^aeiou][aeiou][^aeiou])*', paranoid)
nltk.re_show(r'\w+|[^\w\s]+', paranoid)

{Please} {could} {you} {stop} {the} {noise}?
{I}'{m} {tryna} {get} {some} {rest}
{From} {all} {the} {unborn} {chicken}
{Voices} {in} {my} {head}
{Please} could you stop the noise?
{I}'m tryna get some rest
{From} all the unborn chicken
{Voices} in my head
['pat', 'paut', 'peat', 'pet', 'piet', 'pit', 'poet', 'poot', 'pot', 'pout', 'put']
[]
{}P{}l{}e{}a{se }{}c{}o{}u{}l{}d{} {}y{}o{}u{} {}s{top}{} {}t{he }{}n{}o{}i{se?}{}
{}I{}'{}m{} {}t{}r{}y{na get}{} {som}{}e{} {res}{}t{}
{}F{rom al}{}l{} {}t{he }{}u{}n{bor}{}n{} {}c{hicken}{}
{}V{}o{}i{ces in}{} {}m{}y{} {}h{}e{}a{}d{}
{Please} {could} {you} {stop} {the} {noise}{?}
{I}{'}{m} {tryna} {get} {some} {rest}
{From} {all} {the} {unborn} {chicken}
{Voices} {in} {my} {head}


### 7. Write regular expressions to match the following classes of strings:

1. A single determiner (assume that a, an, and the are the only determiners).
2. An arithmetic expression using integers, addition, and multiplication, such as 2*3+8.

In [136]:
paranoid_full_raw = open('paranoid_full.txt')
paranoid_full = paranoid_full_raw.read()
nltk.re_show(r'(\ban?\b)|(\bthe\b)', paranoid_full)

Please could you stop {the} noise?
I'm tryna get some rest
From all {the} unborn chicken
Voices in my head
What's that? (I may be paranoid, but not {an} android)
What's that? (I may be paranoid, but not {an} android)
When I am king
You will be first against {the} wall
With your opinion
Which is of no consequence at all
What's that? (I may be paranoid, but no android)
What's that? (I may be paranoid, but no android)
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la-la-la-la
La, la
Ambition makes you look pretty ugly
Kicking, squealing, Gucci little piggy
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la
You don't remember, you don't remember
Why don't you remember my name?
Off with his head, man, off with his head, man
Why don't you remember my name?
I guess he does
Ah, oh, oh, oh
Oh, oh, oh, oh
Oh, oh
Oh, oh
Rain down, rain down
Come on, rain down on me
From {a} great height
From {a} great height, height
Rain down, rain down
Come on, rain down on me
From {a} great height
From {

In [137]:
arithmetics_7b = "2+2+2+2+2, 2+2, 5*4, 73-74+73-74, 34/2, 2*3+8, banana, apple, DNA, 451Fahrenheit" # {451}Fahrenheit is picked in both cases

nltk.re_show(r'-?\d+(\s*[\+\*]\s*-?\d+)*', arithmetics_7b) # ChatGPT
nltk.re_show(r'([^\sa-zA-Z,]*[\+\*]*)*[^\s,a-zA-Z]', arithmetics_7b) # my version 

{2+2+2+2+2}, {2+2}, {5*4}, {73}{-74+73}{-74}, {34}/{2}, {2*3+8}, banana, apple, DNA, {451}Fahrenheit
{2+2+2+2+2}, {2+2}, {5*4}, {73-74+73-74}, {34/2}, {2*3+8}, banana, apple, DNA, {451}Fahrenheit


#### 8. Write a utility function that takes a URL as its argument, and returns the contents of the URL, with all HTML markup removed. Use from urllib import request and then request.urlopen('http://nltk.org/').read().decode('utf8') to access the contents of the URL.

In [138]:
def url_opener(link):
    url = link 
    response = request.urlopen(url) 
    raw = response.read().decode('utf8')
    ready_url = BeautifulSoup(raw, 'html.parser').get_text()
    print(ready_url) 
    
url_opener('http://nltk.org/')
url_opener('https://mariaonoeva.github.io/Biennial2024/')







NLTK :: Natural Language Toolkit















NLTK



Documentation














NLTK Documentation

API Reference
Example Usage
Module Index
Wiki
FAQ
Open Issues
NLTK on GitHub

Installation

Installing NLTK
Installing NLTK Data

More

Release Notes
Contributing to NLTK
NLTK Team






Natural Language Toolkit¶
NLTK is a leading platform for building Python programs to work with human language data.
It provides easy-to-use interfaces to over 50 corpora and lexical
resources such as WordNet,
along with a suite of text processing libraries for classification, tokenization, stemming, tagging, parsing, and semantic reasoning,
wrappers for industrial-strength NLP libraries,
and an active discussion forum.
Thanks to a hands-on guide introducing programming fundamentals alongside topics in computational linguistics, plus comprehensive API documentation,
NLTK is suitable for linguists, engineers, students, educators, researchers, and industry users alike.
NLTK is available for Window

#### 9.  Save some text into a file corpus.txt. Define a function load(f) that reads from the file named in its sole argument, and returns a string containing the text of the file.

1. Use nltk.regexp_tokenize() to create a tokenizer that tokenizes the various kinds of punctuation in this text. Use one multi-line regular expression, with inline comments, using the verbose flag (?x).

In [139]:
def load(my_text_raw): 
    my_text = open(my_text_raw)
    my_text_read = my_text.read() 
    return my_text_read

def punct_counter(my_text):
    counter_list = []
    punct = nltk.regexp_tokenize(load(my_text), r'[^\w\s\d]')
    for i in punct: 
        counter_list.append(punct.count(i))
    ready_dict = dict(zip(punct, counter_list))
    sorted_dict = dict(sorted(ready_dict.items(), key=lambda item: item[1], reverse=True))
    return sorted_dict.items()

punct_counter('paranoid_full.txt')
eng_HP = punct_counter('/Users/maria.onoeva/Desktop/new_folder/HP_1.txt')
ru_HP = punct_counter('/Users/maria.onoeva/Desktop/new_folder/HP_1ru.txt')

In [140]:
from tabulate import tabulate 
headers = ['Punctuation', 'Count']

print(tabulate(eng_HP, headers=headers))
print(tabulate(ru_HP, headers=headers))

# I want them side by side 


Punctuation      Count
-------------  -------
.                 6136
,                 5658
"                 4758
'                 3141
-                 1990
?                  754
!                  474
;                  135
:                   69
)                   33
(                   30
*                    2
~                    1
\                    1
Punctuation      Count
-------------  -------
,                10190
.                 6355
—                 4598
-                 1038
…                  805
?                  761
!                  620
«                  222
»                  221
:                  140
*                  117
(                   13
)                   13
;                    3
№                    1
“                    1
”                    1


#### Harry Potter 1 punctuation: Eng and Ru 
##### Eng: 
- why are there 30 of '(' and 33 of ')'? 
- questions: 754
##### Ru: 
- same for Ru for kavychki, but skobki are even
- so many commas??? 
- questions: 761 > 7 more > pretty consistent, I can imagine contexts where several questions mark used to express anger or surprise 

In [141]:
# ChatGPT to display the tables side by side
# Generate table strings
table1 = tabulate(eng_HP, headers=['Punctuation ENG', 'Count'], tablefmt='plain')
table2 = tabulate(ru_HP, headers=['Punctuation RU', 'Count'], tablefmt='plain')

# Split tables into lines
lines1 = table1.split('\n')
lines2 = table2.split('\n')

# Determine the max width of the first table
max_width = max(len(line) for line in lines1)

# Determine the longer table for iteration
longest_table_length = max(len(lines1), len(lines2))

# Print tables side by side, handling uneven number of rows
for i in range(longest_table_length):
    line1 = lines1[i] if i < len(lines1) else ""
    line2 = lines2[i] if i < len(lines2) else ""
    print(f"{line1.ljust(max_width)}    {line2}")

Punctuation ENG      Count    Punctuation RU      Count
.                     6136    ,                   10190
,                     5658    .                    6355
"                     4758    —                    4598
'                     3141    -                    1038
-                     1990    …                     805
?                      754    ?                     761
!                      474    !                     620
;                      135    «                     222
:                       69    »                     221
)                       33    :                     140
(                       30    *                     117
*                        2    (                      13
~                        1    )                      13
\                        1    ;                       3
                              №                       1
                              “                       1
                              ”                 

#### 9. Save some text into a file corpus.txt. Define a function load(f) that reads from the file named in its sole argument, and returns a string containing the text of the file.

2. Use nltk.regexp_tokenize() to create a tokenizer that tokenizes the following kinds of expression: monetary amounts; dates; names of people and organizations.

In [142]:
# monetary amounts? 
monetary_amounts = "200$, 363524$, 587$, 65€, 8746CZK, 8₽, 9382₽, 87362513537CZK, Grand Theft Auto V, Corpus paper, April, Thursday, 76F, 17C"
money_pattern = r'\d+([\$€₽]|CZK)+' 

nltk.re_show(money_pattern, monetary_amounts)

{200$}, {363524$}, {587$}, {65€}, {8746CZK}, {8₽}, {9382₽}, {87362513537CZK}, Grand Theft Auto V, Corpus paper, April, Thursday, 76F, 17C


In [146]:
# dates
dates_more = '2024-04-04, 2024-04-05, Your Top Songs 2023, 2024-04-06, pinguin, 2024-04-07, 2024-04-08, Ice-cream, 2024-04-09, 2024-04-10, Sněžka'
dates_pattern = r'\d{4}-\d{2}-\d{2}'
nltk.re_show(dates_pattern, dates_more)

# names of people and organizations
names_pattern = r'[A-Z]+'
nltk.re_show(names_pattern, dates_more)

{2024-04-04}, {2024-04-05}, Your Top Songs 2023, {2024-04-06}, pinguin, {2024-04-07}, {2024-04-08}, Ice-cream, {2024-04-09}, {2024-04-10}, Sněžka
2024-04-04, 2024-04-05, {Y}our {T}op {S}ongs 2023, 2024-04-06, pinguin, 2024-04-07, 2024-04-08, {I}ce-cream, 2024-04-09, 2024-04-10, {S}něžka
